File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(bf.readLine());
        ingredients = new int[n][2];
        for (int i = 0; i < n; i++) {
            String[] split = bf.readLine().split(" ");
            ingredients[i][0] = Integer.parseInt(split[0]);
            ingredients[i][1] = Integer.parseInt(split[1]);
        }
        recursive(0, 1, 0, 0);
        System.out.println(answer);
    }


    static int[][] ingredients;
    static int answer = Integer.MAX_VALUE;

    static private void recursive(int index, int sour, int bitter, int use) {
        if (index == ingredients.length) {
            if (use == 0) {
                return;
            }
            answer = Math.min(answer, Math.abs(sour - bitter));
            return;
        }

        recursive(index + 1, sour * ingredients[index][0], bitter + ingredients[index][1], use + 1);
        recursive(index + 1, sour, bitter, use);
    }
}

File: 백준/Silver/2961. 도영이가 만든 맛있는 음식/도영이가 만든 맛있는 음식.py
from sys import stdin
input = stdin.readline

def solution(depth:int, used:int, S:int, B:int):
    if depth == N:
        if used != 0:
            results.append(abs(S-B))
        return
    # 사용하거나
    solution(depth+1, used+1, S*SB[depth][0], B+SB[depth][1])
    # 사용하지 않거나
    solution(depth+1, used, S, B)

N = int(input())
SB = [list(map(int, input().split())) for _ in range(N)]
results = []

solution(0,0,1,0)
print(min(results))